crates/storage/src/connection/trie.rs

```diff
@@ -222,33 +250,71 @@ impl Transaction<'_> {
         self.trie_node_hash(index, "trie_storage")
     }
 
-    fn delete_contract_roots(&self, contract: ContractAddress) -> anyhow::Result<()> {
-        let mut stmt = self
-            .inner()
-            .prepare_cached("DELETE FROM contract_roots WHERE contract_address = ?")?;
-        stmt.execute(params![&contract])?;
+    fn delete_contract_roots(
+        &self,
+        contract: ContractAddress,
+        before_block: BlockNumber,
+    ) -> anyhow::Result<()> {
+        let mut stmt = self.inner().prepare_cached(
+            "DELETE FROM contract_roots WHERE contract_address = ? AND block_number < ?",
+        )?;
+        stmt.execute(params![&contract, &before_block])?;
         Ok(())
     }
 
-    fn remove_trie(&self, removed: &[u64], table: &'static str) -> anyhow::Result<()> {
-        let mut stmt = self
+    fn remove_trie(
+        &self,
+        removed: &[u64],
+        block_number: BlockNumber,
+        num_blocks_kept: u64,
+        table: &'static str,
+    ) -> anyhow::Result<()> {
+        // Delete nodes marked as ready for deletion.
+        let num_removed = self
             .inner()
-            .prepare_cached(&format!("DELETE FROM {table} WHERE idx = ?"))
-            .context("Creating delete statement")?;
+            .execute(
+                &format!(
+                    r"DELETE FROM {table} WHERE EXISTS (
+                        SELECT 1 FROM {table}_removals WHERE idx = {table}.idx AND block_number <= ?
+                    )"
+                ),
+                params![&block_number],
+            )
+            .context("Deleting nodes")?;
+        metrics::counter!(METRIC_TRIE_NODES_REMOVED, num_removed.try_into().unwrap(), "table" => table);
 
-        let number_of_nodes_removed = removed.len() as u64;
-        metrics::counter!(METRIC_TRIE_NODES_REMOVED, number_of_nodes_removed, "table" => table);
+        // Delete the removal markers.
+        self.inner()
+            .execute(
+                &format!(r"DELETE FROM {table}_removals WHERE block_number <= ?"),
+                params![&block_number],
+            )
+            .context("Deleting nodes")?;
 
-        for idx in removed {
-            stmt.execute(params![idx]).context("Deleting node")?;
+        // Mark the input nodes as ready for deletion at block `block_number + num_blocks_kept`.
+        let mut stmt = self
+            .inner()
+            .prepare(&format!(
+                r"INSERT INTO {table}_removals (idx, block_number) VALUES (?, ?)"
+            ))
+            .context("Creating removal statement")?;
+        for &idx in removed {
+            stmt.execute(params![&idx, &(block_number + num_blocks_kept)])
```

Comment:
Hmm, so in the `_removal` table we're storing the number of the block at which we'll actually delete the nodes?

That might lead to slightly counter-intuitive behavior. If I change the `num_blocks_kept` setting on the command line then I'd expect the setting to take effect for blocks already in the DB but this way the setting will only take effect for new blocks synced.

Reply:
Plus one - we should store the current block number i.e. the number at which it safe to remove these.